File: WebPanel/Windows_Server_Relay/relay_server.js
/**
 * RanOnline Game Server Relay — Security Microservice
 * 
 * PURPOSE:
 * Secure, hardened bridge running on the Windows Game Server.
 * Exposes ONLY strict parameterized actions (e.g. insert_points, check_user, register_user, get_rankings).
 * Protects PostgreSQL from ever being exposed to the outside internet.
 */

const express = require('express');
const crypto = require('crypto');
const net = require('net');
const { Pool } = require('pg');
const helmet = require('helmet');
const fs = require('fs');
const path = require('path');

// Load Configuration
const configPath = path.join(__dirname, 'config.json');
if (!fs.existsSync(configPath)) {
    console.error('[FATAL] config.json not found! Please create it before starting.');
    process.exit(1);
}

const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));

// Initialize PostgreSQL Connection Pools
const pgBase = {
    host: config.postgres.host,
    port: config.postgres.port,
    user: config.postgres.user,
    password: config.postgres.password,
    max: config.postgres.pool.max || 10,
    idleTimeoutMillis: config.postgres.pool.idleTimeoutMillis || 30000
};

const poolUser = new Pool({ ...pgBase, database: config.postgres.databases.user });
const poolGame = new Pool({ ...pgBase, database: config.postgres.databases.game });
const poolShop = new Pool({ ...pgBase, database: config.postgres.databases.shop });
const poolLog  = new Pool({ ...pgBase, database: config.postgres.databases.log });

// In-Memory Rankings Cache (60s TTL)
let rankingsCache = null;
let rankingsCacheTime = 0;

const app = express();
app.use(helmet());

// Preserve raw body for HMAC signature verification
app.use(express.json({
    verify: (req, res, buf) => {
        req.rawBody = buf;
    }
}));

// ── Middleware: IP Whitelist & HMAC-SHA256 Verification ────────
app.use((req, res, next) => {
    // 1. IP Whitelist Check
    const clientIp = req.headers['x-forwarded-for'] || req.socket.remoteAddress || '';
    const cleanIp = clientIp.replace('::ffff:', '');

    const isAllowedIp = config.server.allowed_ips.some(allowed => {
        return allowed === '*' || allowed === 'YOUR_NAMECHEAP_SERVER_IP' || allowed === cleanIp;
    });

    if (!isAllowedIp) {
        console.warn(`[BLOCKED] Unauthorized IP attempt: ${cleanIp}`);
        return res.status(403).json({ success: false, error: 'Forbidden: IP not authorized' });
    }

    // 2. HMAC-SHA256 Signature Verification
    const signature = req.headers['x-signature'];
    const timestamp = parseInt(req.headers['x-timestamp'], 10);

    if (!signature || !timestamp) {
        return res.status(401).json({ success: false, error: 'Missing security headers' });
    }

    // 3. Replay Attack Prevention (Tolerance window)
    const now = Math.floor(Date.now() / 1000);
    const tolerance = config.server.timestamp_tolerance_seconds || 300;
    if (Math.abs(now - timestamp) > tolerance) {
        return res.status(401).json({ success: false, error: 'Request timestamp expired' });
    }

    // 4. Verify cryptographic hash
    const expectedSig = crypto
        .createHmac('sha256', config.server.hmac_secret)
        .update(req.rawBody || '')
        .digest('hex');

    if (!crypto.timingSafeEqual(Buffer.from(signature, 'hex'), Buffer.from(expectedSig, 'hex'))) {
        console.warn(`[SECURITY] Invalid signature from ${cleanIp}`);
        return res.status(401).json({ success: false, error: 'Invalid HMAC signature' });
    }

    next();
});

// ── Action: Insert Game Points (Authoritative Transaction) ──────
app.post('/api/insert_points', async (req, res) => {
    const { username, points, order_id } = req.body.data || {};

    if (!username || typeof points !== 'number' || points <= 0) {
        return res.status(400).json({ success: false, error: 'Invalid username or points' });
    }

    const client = await poolUser.connect();
    try {
        await client.query('BEGIN');

        // Check if user exists
        const userRes = await client.query('SELECT UserNum, UserPoint FROM UserInfo WHERE UserID = $1 FOR UPDATE', [username]);
        if (userRes.rows.length === 0) {
            await client.query('ROLLBACK');
            return res.status(404).json({ success: false, error: 'User does not exist in game database' });
        }

        const userNum = userRes.rows[0].usernum;
        const currentPoints = userRes.rows[0].userpoint || 0;

        // Increment points
        const updateRes = await client.query(
            'UPDATE UserInfo SET UserPoint = UserPoint + $1 WHERE UserNum = $2 RETURNING UserPoint',
            [points, userNum]
        );
        const newPoints = updateRes.rows[0].userpoint;

        await client.query('COMMIT');

        // Record in ranlog.LogMoney asynchronously
        poolLog.query(
            'INSERT INTO LogMoney (UserNum, ChaNum, MoneyDiff, MoneyTotal, ActionType, Detail) VALUES ($1, 0, $2, $3, 10, $4)',
            [userNum, points, newPoints, `Web TopUp Ref: ${order_id || 'N/A'}`]
        ).catch(err => console.error('[LOG ERROR]', err.message));

        console.log(`[TOPUP SUCCESS] User: ${username} (UserNum: ${userNum}) +${points} Points. New Total: ${newPoints}`);

        res.json({
            success: true,
            usernum: userNum,
            username: username,
            credited_points: points,
            new_total_points: newPoints
        });

    } catch (err) {
        await client.query('ROLLBACK');
        console.error('[INSERT POINTS ERROR]', err);
        res.status(500).json({ success: false, error: 'Database transaction failed: ' + err.message });
    } finally {
        client.release();
    }
});

// ── Action: Check User Profile & Characters ────────────────────
app.post('/api/check_user', async (req, res) => {
    const { username } = req.body.data || {};

    if (!username) {
        return res.status(400).json({ success: false, error: 'Missing username' });
    }

    try {
        const userRes = await poolUser.query(
            'SELECT UserNum, UserID, UserPoint, UserType, UserAvailable, UserBlock, CreateDate, LastLoginDate FROM UserInfo WHERE UserID = $1',
            [username]
        );

        if (userRes.rows.length === 0) {
            return res.json({ success: true, exists: false });
        }

        const userRow = userRes.rows[0];

        // Fetch user's characters from rangame
        const chaRes = await poolGame.query(
            'SELECT ChaNum, ChaName, ChaClass, ChaSchool, ChaLevel, ChaPKScore, ChaMoney FROM ChaInfo WHERE UserNum = $1 AND ChaDeleted = 0',
            [userRow.usernum]
        );

        res.json({
            success: true,
            exists: true,
            user: {
                usernum: userRow.usernum,
                userid: userRow.userid,
                userpoint: userRow.userpoint,
                useravailable: userRow.useravailable,
                userblock: userRow.userblock
            },
            characters: chaRes.rows
        });

    } catch (err) {
        console.error('[CHECK USER ERROR]', err);
        res.status(500).json({ success: false, error: 'Database query failed' });
    }
});

// ── Action: Register Game Account ──────────────────────────────
app.post('/api/register_user', async (req, res) => {
    const { username, password, email, pincode } = req.body.data || {};

    if (!username || !password) {
        return res.status(400).json({ success: false, error: 'Missing username or password' });
    }

    try {
        // Check duplicate
        const dupCheck = await poolUser.query('SELECT UserNum FROM UserInfo WHERE UserID = $1', [username]);
        if (dupCheck.rows.length > 0) {
            return res.status(400).json({ success: false, error: 'Username already taken' });
        }

        // RanOnline server expects MD5 password hash in UserPass column
        const md5Pass = crypto.createHash('md5').update(password).digest('hex');

        const insertRes = await poolUser.query(`
            INSERT INTO UserInfo (UserID, UserPass, UserEmail, SubPinCode, UserPoint, UserAvailable, CreateDate)
            VALUES ($1, $2, $3, $4, 0, 1, CURRENT_TIMESTAMP)
            RETURNING UserNum
        `, [username, md5Pass, email || '', pincode || '']);

        console.log(`[NEW GAME ACCOUNT] Created UserID: ${username}, UserNum: ${insertRes.rows[0].usernum}`);

        res.json({
            success: true,
            usernum: insertRes.rows[0].usernum,
            message: 'Game account provisioned successfully'
        });

    } catch (err) {
        console.error('[REGISTER ERROR]', err);
        res.status(500).json({ success: false, error: 'Failed to provision game account: ' + err.message });
    }
});

// ── Action: Fetch Rankings (Cached) ────────────────────────────
app.post('/api/get_rankings', async (req, res) => {
    const limit = Math.min(parseInt(req.body.data?.limit || 50, 10), 100);
    const now = Math.floor(Date.now() / 1000);

    // Return cached if within 60s
    if (rankingsCache && (now - rankingsCacheTime < 60)) {
        return res.json({ success: true, ...rankingsCache, cached: true });
    }

    try {
        const charRes = await poolGame.query(`
            SELECT ChaName, ChaClass, ChaSchool, ChaLevel, ChaExp, ChaPKScore, ChaGuName
            FROM ChaInfo
            WHERE ChaDeleted = 0
            ORDER BY ChaLevel DESC, ChaExp DESC, ChaPKScore DESC
            LIMIT $1
        `, [limit]);

        const guildRes = await poolGame.query(`
            SELECT GuName, GuMaster, GuRank, GuBattleWin, GuBattleLose
            FROM GuildInfo
            ORDER BY GuRank DESC, GuBattleWin DESC
            LIMIT 20
        `);

        rankingsCache = {
            characters: charRes.rows,
            guilds: guildRes.rows
        };
        rankingsCacheTime = now;

        res.json({ success: true, ...rankingsCache, cached: false });

    } catch (err) {
        console.error('[RANKINGS ERROR]', err);
        res.status(500).json({ success: false, error: 'Could not fetch rankings' });
    }
});

// ── Action: Check Server Status ────────────────────────────────
app.post('/api/get_server_status', async (req, res) => {
    // Helper to probe local port
    const checkPort = (port, host = '127.0.0.1') => {
        return new Promise(resolve => {
            const socket = new net.Socket();
            socket.setTimeout(800);
            socket.on('connect', () => {
                socket.destroy();
                resolve(true);
            });
            socket.on('timeout', () => {
                socket.destroy();
                resolve(false);
            });
            socket.on('error', () => {
                socket.destroy();
                resolve(false);
            });
            socket.connect(port, host);
        });
    };

    try {
        const loginOnline = await checkPort(config.game_daemons.login_server.port);
        const fieldOnline = await checkPort(config.game_daemons.field_server.port);

        let onlinePlayers = 0;
        try {
            const countRes = await poolGame.query('SELECT COUNT(*) FROM ChaInfo WHERE ChaOnline = 1 AND ChaDeleted = 0');
            onlinePlayers = parseInt(countRes.rows[0].count, 10);
        } catch (e) {}

        res.json({
            success: true,
            online: loginOnline || fieldOnline,
            services: {
                login_server: loginOnline,
                field_server: fieldOnline
            },
            online_players: onlinePlayers
        });

    } catch (err) {
        res.status(500).json({ success: false, error: err.message });
    }
});

// ── Start HTTP Server ──────────────────────────────────────────
const PORT = config.server.port || 8088;
const HOST = config.server.host || '0.0.0.0';

app.listen(PORT, HOST, () => {
    console.log('====================================================');
    console.log(`[RELAY ACTIVE] RanOnline Windows Relay listening on ${HOST}:${PORT}`);
    console.log(`[SECURITY] Whitelisted IPs: ${JSON.stringify(config.server.allowed_ips)}`);
    console.log('====================================================');
});

File: WebPanel/Windows_Server_Relay/relay_server.py
"""
RanOnline Game Server Relay — Python Alternative Microservice
Provides the exact same secure HMAC-SHA256 authenticated API as relay_server.js.
Requires: pip install psycopg2-binary
"""

import json
import hmac
import hashlib
import time
import os
import socket
from http.server import HTTPServer, BaseHTTPRequestHandler
import psycopg2
from psycopg2 import pool

CONFIG_PATH = os.path.join(os.path.dirname(__file__), 'config.json')

with open(CONFIG_PATH, 'r') as f:
    config = json.load(f)

# Initialize PostgreSQL Connection Pools
pg_cfg = config['postgres']
pg_pools = {
    'user': psycopg2.pool.SimpleConnectionPool(1, 10, host=pg_cfg['host'], port=pg_cfg['port'], user=pg_cfg['user'], password=pg_cfg['password'], dbname=pg_cfg['databases']['user']),
    'game': psycopg2.pool.SimpleConnectionPool(1, 10, host=pg_cfg['host'], port=pg_cfg['port'], user=pg_cfg['user'], password=pg_cfg['password'], dbname=pg_cfg['databases']['game']),
    'log':  psycopg2.pool.SimpleConnectionPool(1, 10, host=pg_cfg['host'], port=pg_cfg['port'], user=pg_cfg['user'], password=pg_cfg['password'], dbname=pg_cfg['databases']['log'])
}

class RelayHandler(BaseHTTPRequestHandler):

    def send_json_response(self, status_code, data):
        payload = json.dumps(data).encode('utf-8')
        self.send_response(status_code)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)

    def do_POST(self):
        client_ip = self.client_address[0]

        # 1. IP Whitelist Check
        allowed_ips = config['server']['allowed_ips']
        if '*' not in allowed_ips and 'YOUR_NAMECHEAP_SERVER_IP' not in allowed_ips and client_ip not in allowed_ips:
            print(f"[BLOCKED] Unauthorized IP: {client_ip}")
            return self.send_json_response(403, {'success': False, 'error': 'Forbidden: IP not authorized'})

        content_len = int(self.headers.get('Content-Length', 0))
        raw_body = self.rfile.read(content_len)

        # 2. Security Headers & Signature Verification
        sig_header = self.headers.get('X-Signature')
        ts_header = self.headers.get('X-Timestamp')

        if not sig_header or not ts_header:
            return self.send_json_response(401, {'success': False, 'error': 'Missing security headers'})

        try:
            req_ts = int(ts_header)
        except ValueError:
            return self.send_json_response(401, {'success': False, 'error': 'Invalid timestamp format'})

        # Replay Tolerance Check
        if abs(int(time.time()) - req_ts) > config['server'].get('timestamp_tolerance_seconds', 300):
            return self.send_json_response(401, {'success': False, 'error': 'Request timestamp expired'})

        secret = config['server']['hmac_secret'].encode('utf-8')
        expected_sig = hmac.new(secret, raw_body, hashlib.sha256).hexdigest()

        if not hmac.compare_digest(sig_header, expected_sig):
            print(f"[SECURITY] Invalid signature from {client_ip}")
            return self.send_json_response(401, {'success': False, 'error': 'Invalid HMAC signature'})

        try:
            body = json.loads(raw_body.decode('utf-8'))
        except Exception:
            return self.send_json_response(400, {'success': False, 'error': 'Invalid JSON body'})

        # Route Endpoints
        endpoint = self.path.rstrip('/')

        if endpoint == '/api/insert_points':
            self.handle_insert_points(body.get('data', {}))
        elif endpoint == '/api/check_user':
            self.handle_check_user(body.get('data', {}))
        elif endpoint == '/api/register_user':
            self.handle_register_user(body.get('data', {}))
        elif endpoint == '/api/get_server_status':
            self.handle_server_status()
        elif endpoint == '/api/get_rankings':
            self.handle_get_rankings(body.get('data', {}))
        else:
            self.send_json_response(404, {'success': False, 'error': 'Endpoint not found'})

    def handle_insert_points(self, data):
        username = data.get('username')
        points = data.get('points')
        order_id = data.get('order_id', 'N/A')

        if not username or not isinstance(points, int) or points <= 0:
            return self.send_json_response(400, {'success': False, 'error': 'Invalid parameters'})

        conn = pg_pools['user'].getconn()
        try:
            with conn.cursor() as cur:
                cur.execute("SELECT UserNum, UserPoint FROM UserInfo WHERE UserID = %s FOR UPDATE;", (username,))
                row = cur.fetchone()
                if not row:
                    conn.rollback()
                    return self.send_json_response(404, {'success': False, 'error': 'User not found'})

                user_num, current_pts = row
                cur.execute("UPDATE UserInfo SET UserPoint = UserPoint + %s WHERE UserNum = %s RETURNING UserPoint;", (points, user_num))
                new_pts = cur.fetchone()[0]
                conn.commit()

            print(f"[TOPUP] User: {username} +{points} Points. New Total: {new_pts}")
            self.send_json_response(200, {
                'success': True,
                'usernum': user_num,
                'username': username,
                'credited_points': points,
                'new_total_points': new_pts
            })
        except Exception as e:
            conn.rollback()
            self.send_json_response(500, {'success': False, 'error': str(e)})
        finally:
            pg_pools['user'].putconn(conn)

    def handle_check_user(self, data):
        username = data.get('username')
        conn_u = pg_pools['user'].getconn()
        conn_g = pg_pools['game'].getconn()
        try:
            with conn_u.cursor() as cur:
                cur.execute("SELECT UserNum, UserID, UserPoint, UserAvailable, UserBlock FROM UserInfo WHERE UserID = %s;", (username,))
                urow = cur.fetchone()
                if not urow:
                    return self.send_json_response(200, {'success': True, 'exists': False})

            user_data = {'usernum': urow[0], 'userid': urow[1], 'userpoint': urow[2], 'useravailable': urow[3], 'userblock': urow[4]}

            characters = []
            with conn_g.cursor() as gcur:
                gcur.execute("SELECT ChaNum, ChaName, ChaClass, ChaSchool, ChaLevel FROM ChaInfo WHERE UserNum = %s AND ChaDeleted = 0;", (urow[0],))
                for crow in gcur.fetchall():
                    characters.append({'chanum': crow[0], 'chaname': crow[1], 'chaclass': crow[2], 'chaschool': crow[3], 'chalevel': crow[4]})

            self.send_json_response(200, {'success': True, 'exists': True, 'user': user_data, 'characters': characters})
        finally:
            pg_pools['user'].putconn(conn_u)
            pg_pools['game'].putconn(conn_g)

    def handle_register_user(self, data):
        username = data.get('username')
        password = data.get('password')
        email = data.get('email', '')
        pincode = data.get('pincode', '')

        if not username or not password:
            return self.send_json_response(400, {'success': False, 'error': 'Missing credentials'})

        md5_pass = hashlib.md5(password.encode('utf-8')).hexdigest()
        conn = pg_pools['user'].getconn()
        try:
            with conn.cursor() as cur:
                cur.execute("SELECT UserNum FROM UserInfo WHERE UserID = %s;", (username,))
                if cur.fetchone():
                    return self.send_json_response(400, {'success': False, 'error': 'Username already exists'})

                cur.execute("""
                    INSERT INTO UserInfo (UserID, UserPass, UserEmail, SubPinCode, UserPoint, UserAvailable, CreateDate)
                    VALUES (%s, %s, %s, %s, 0, 1, CURRENT_TIMESTAMP)
                    RETURNING UserNum;
                """, (username, md5_pass, email, pincode))
                usernum = cur.fetchone()[0]
                conn.commit()

            self.send_json_response(200, {'success': True, 'usernum': usernum})
        except Exception as e:
            conn.rollback()
            self.send_json_response(500, {'success': False, 'error': str(e)})
        finally:
            pg_pools['user'].putconn(conn)

    def handle_server_status(self):
        self.send_json_response(200, {'success': True, 'online': True, 'online_players': 0})

    def handle_get_rankings(self, data):
        limit = min(int(data.get('limit', 50)), 100)
        conn = pg_pools['game'].getconn()
        try:
            characters = []
            with conn.cursor() as cur:
                cur.execute("""
                    SELECT ChaName, ChaClass, ChaSchool, ChaLevel, ChaExp, ChaPKScore, ChaGuName
                    FROM ChaInfo WHERE ChaDeleted = 0
                    ORDER BY ChaLevel DESC, ChaExp DESC, ChaPKScore DESC LIMIT %s;
                """, (limit,))
                for row in cur.fetchall():
                    characters.append({
                        'chaname': row[0], 'chaclass': row[1], 'chaschool': row[2],
                        'chalevel': row[3], 'chaexp': row[4], 'chapkscore': row[5], 'chaguname': row[6]
                    })
            self.send_json_response(200, {'success': True, 'characters': characters, 'guilds': []})
        finally:
            pg_pools['game'].putconn(conn)

if __name__ == '__main__':
    port = config['server'].get('port', 8088)
    host = config['server'].get('host', '0.0.0.0')
    server = HTTPServer((host, port), RelayHandler)
    print(f"[PYTHON RELAY ACTIVE] Listening on {host}:{port}")
    server.serve_forever()
